python/ray/train/lightning/_lightning_utils.py

```diff
@@ -194,17 +222,23 @@ def _session_report(self, trainer: "pl.Trainer", stage: str):
             if isinstance(v, torch.Tensor):
                 metrics[k] = v.item()
 
-        # Report latest saved checkpoint
-        # Note that AIR only takes the checkpoint of rank 0.
-        # Save a dummy checkpoint on the other workers to avoid blocking.
+        # Ensures all workers already finish writing their checkpoints.
+        trainer.strategy.barrier()
+
+        # Create and report the latest checkpoint
         with tempfile.TemporaryDirectory() as tmpdir:
-            if trainer.global_rank == 0:
-                shutil.copy(self.last_model_path, f"{tmpdir}/{MODEL_KEY}")
-                checkpoint = LightningCheckpoint.from_directory(path=tmpdir)
-            else:
-                checkpoint = LightningCheckpoint.from_dict(
-                    {"rank": session.get_world_rank()}
-                )
+            src_model_path = os.path.expanduser(self.last_model_path)
+            dst_model_path = f"{tmpdir}/{MODEL_KEY}"
+
+            if self.is_report_rank:
+                if os.path.isdir(src_model_path):
+                    shutil.copytree(src_model_path, dst_model_path)
+                elif os.path.isfile(src_model_path):
+                    shutil.copy(src_model_path, dst_model_path)
+
+            # Only the report_rank worker creates the actual checkpoints.
```

Comment:
Can you add some more description here to describe the behavior for ddp vs. deepspeed?

Something like:

```
For DDP: Each node has the same set of weights, so only the global rank 0 worker needs to report.
For deepspeed: Each node has a unique set of weights, so the local rank 0 workers report the checkpoint shards for all workers on its node. 
```

Reply:
ok